### PR TITLE
simd-doc: allow for continued parsing after errors

### LINKED ISSUE #1584: simd-doc: handle parser re-entrancy on parsing errors
It should be possible to start parsing at an arbitrary journal offset, encounter a JSON decoding error (because the offset is not a document boundary), but to re-poll the Stream and continue parsing.

### PR DESCRIPTION
**Pull request by jgraettinger**
Update Parser to support continued parsing after an error in parsing a JSON document. Use a polling-based design, where a caller can add one or more chunk()'s to a Parser and then repeatedly poll it to drain all parsed or transcoded documents and errors.

Also update error tracking to provide fine-grain localization of exactly where, in the input stream, the erroring document is located.

Enhance fuzz tests to mix in errors alongside valid documents, and verify that parsing and transcoding give exactly the same outcomes.

Resolves #1584

**Workflow steps:**

(How does one use this feature, and how has it changed)

**Documentation links affected:**

(list any [documentation links](https://docs.google.com/document/d/1SRC9VS9zyCzWl3n4HXHbc4wPB1eLxJHkA2rtu9ZNokM/edit?usp=sharing) that you created, or existing ones that you've identified as needing updates, along with a brief description)

**Notes for reviewers:**

(anything that might help someone review this PR)

- - -
This change is [Reviewable](https://reviewable.io/reviews/estuary/flow/1589)

### CODE REVIEW

**Review by jshearer — APPROVED** [review]
Couple small questions but LGTM -- once merged, I'll update Dekaf to use this instead of its own skip-to-next-document logic

File: crates/simd-doc/src/tests/fixtures.rs
Comment (jgraettinger) [reply crates/simd-doc/src/tests/fixtures.rs:67]: yes. The guarantee we provide for parse_many() or transcode_many() is that we handle newline-delimited JSON with no interior newlines. That's required because, without this restriction, we don't know where to resume from upon a JSON decoding error.

parse_one() continues to handle interior newlines.

File: crates/simd-doc/src/lib.rs
Comment (jgraettinger) [reply crates/simd-doc/src/lib.rs:70]: It's not at all required for correctness.

**Pull request merged**
